auth/token_issuer_test.go

```diff
@@ -93,25 +94,76 @@ func TestTokenIssuer(t *testing.T) {
 func TestCreateToken(t *testing.T) {
 	e := &ldap.Entry{
 		DN: "some-dn",
+		Attributes: []*ldap.EntryAttribute{
+			{
+				Name:   "uid",
+				Values: []string{"username"},
+			},
+			{
+				Name:   "mail",
+				Values: []string{"username@example.com"},
+			},
+			{
+				Name: "memberOf",
+				Values: []string{
+					"cn=sg-grp1,ou=ORG,ou=Groups,dc=lab,dc=example,dc=com",
+					"CN=sg-grp2,ou=ORG,ou=Groups,dc=lab,dc=example,dc=com",
+					"CN=sg-grp1,ou=ORG2,ou=Groups,dc=lab,dc=example,dc=com",
+				},
+			},
+		},
 	}
 
-	lti := LDAPTokenIssuer{
-		LDAPServer: "some-ldap-server",
-	}
-
-	expectedAssertions := map[string]string{
-		"ldapServer": lti.LDAPServer,
-		"userDN":     e.DN,
+	cases := []struct {
+		name               string
+		tokenIssuer        LDAPTokenIssuer
+		expectedAssertions map[string]string
+		expectedUsername   string
+		expectedGroups     []string
+	}{
+		{
+			name: "get mail as username attribute",
+			tokenIssuer: LDAPTokenIssuer{
+				LDAPServer:        "some-ldap-server",
+				UsernameAttribute: "mail",
+			},
+			expectedAssertions: map[string]string{
+				"ldapServer": "some-ldap-server",
+				"userDN":     e.DN,
+			},
+			expectedUsername: "username@example.com",
+			expectedGroups: []string{
+				"sg-grp1",
+				"sg-grp2",
+			},
+		},
+		{
+			name: "verify backward compatibility",
+			tokenIssuer: LDAPTokenIssuer{
+				LDAPServer: "some-ldap-server",
+			},
+			expectedAssertions: map[string]string{
+				"ldapServer": "some-ldap-server",
+				"userDN":     e.DN,
+			},
+			expectedUsername: e.DN,
+			expectedGroups: []string{
+				"sg-grp1",
+				"sg-grp2",
+			},
+		},
 	}
 
-	tok := lti.createToken(e)
-	if tok.Username != e.DN {
-		t.Errorf("Unexpected username in token. Expected: '%s'. Got: '%s'.", e.DN, tok.Username)
-	}
+	for _, testcase := range cases {
+		tok := testcase.tokenIssuer.createToken(e)
+		if tok.Username != testcase.expectedUsername {
+			t.Errorf("Unexpected username in token. Expected: '%s'. Got: '%s'.", testcase.expectedUsername, tok.Username)
+		}
```

Comment:
Take a look at github.com/stretchr/testify/assert. It cuts down on the boilerplate.

Reply:
yes for new projects we are using that. i will refactor this soon. I am also working on adding metrics to this project.